### PR TITLE
Issue #447 dirty state on replace

### PR DESCRIPTION
**Pull request by GMartigny**
The dirty state works better.

Unit tests fail on local. I can't get why, I'm missing some insight.

### CODE REVIEW

File: src/js/service/HistoryService.js [hunk 1]
```diff
@@ -40,42 +40,89 @@
...
+    if (this.currentUUID) {
+      return this.stateQueue[this.currentUUID];
+    } else {
+      return false;
+    }
+  };
+
+  ns.HistoryService.prototype.pushNewState_ = function(state) {
+    // Generate a random UUID (~1e28 combinations)
+    var uuid = 'xxxxxx'.replace(/x/g, function() {
```
Comment (juliandescottes) [inline src/js/service/HistoryService.js]: Let's reuse pskl.utils.Uuid.generate() here.

File: src/js/service/SavedStatusService.js [hunk 2]
```diff
@@ -1,45 +1,37 @@
...
-  ns.SavedStatusService.prototype.onPiskelReset = function () {
-    var piskel = this.piskelController.getPiskel();
-    // A first PISKEL_RESET is triggered during the load of a new Piskel, it should be ignored
-    // putting a firstResetDone flag as a nasty workaround for this
-    if (piskel.firstResetDone_) {
-      this.updateDirtyStatus(true);
-    } else {
-      piskel.firstResetDone_ = true;
-    }
+  ns.SavedStatusService.prototype.onToolReleased = function () {
```
Comment (juliandescottes) [inline src/js/service/SavedStatusService.js]: Let's remove this method and bind TOOL_RELEASED to updateDirtyStatus directly

File: src/js/service/SavedStatusService.js [hunk 3]
```diff
@@ -1,45 +1,37 @@
...
+    this.updateDirtyStatus();
   };
 
-  ns.SavedStatusService.prototype.onToolReleased = function () {
-    this.updateDirtyStatus(true);
+  ns.SavedStatusService.prototype.onPiskelReset = function () {
```
Comment (juliandescottes) [inline src/js/service/SavedStatusService.js]: Same, let's remove this method and bind PISKEL_RESET to updateDirtyStatus directly

File: src/js/service/SavedStatusService.js [hunk 4]
```diff
@@ -1,45 +1,37 @@
...
+    this.updateDirtyStatus();
   };
 
   ns.SavedStatusService.prototype.onPiskelSaved = function () {
-    this.updateDirtyStatus(false);
+    this.lastSavedStateIndex = this.historyService.getCurrentStateIndex();
+    $.publish(Events.PISKEL_SAVED_STATUS_UPDATE);
```
Comment (juliandescottes) [inline src/js/service/SavedStatusService.js]: use updateDirtyStatus();

File: src/js/service/SavedStatusService.js [hunk 5]
```diff
@@ -1,45 +1,37 @@
...
   };
 
-  ns.SavedStatusService.prototype.updateDirtyStatus = function (status) {
-    var piskel = this.piskelController.getPiskel();
-    if (piskel.isDirty_ != status) {
-      piskel.isDirty_ = status;
-      $.publish(Events.PISKEL_SAVED_STATUS_UPDATE);
-    }
+  ns.SavedStatusService.prototype.updateDirtyStatus = function () {
```
Comment (juliandescottes) [inline src/js/service/SavedStatusService.js]: rename to publishStatusUpdateEvent_ or something like that ?

**Comment by juliandescottes** [comment]
Thanks for working on this!

The changes in SavedStatusService.js and app.js look very good.

For the HistoryService, I had something simpler in mind. What about creating a UUID when we create a state object, add it to the state object. I might be missing something but I don't see why we should refactor the internals of the HistoryService here. Let's discuss!

**Comment by juliandescottes** [comment]
@GMartigny let me know if you still want to work on this. Otherwise, I might pick it up :)

**Comment by GMartigny** [comment]
@juliandescottes No problem, you can take it, it's up to you. I don't have much time currently.

**Comment by juliandescottes** [comment]
Merged in #471

**Pull request closed without merging**